rfc/rfc-0000-k8s-events.md

```diff
@@ -0,0 +1,176 @@
+
+[how-it-works]: #how-it-works
+
+To avoid over-communicating, and making the view of event's overwhelming and dilute for user's we will start with two
+key guiding principles:
+
+1. > Let it be something that isn't satisfactorily presented in other user views (eg: `k get workload`).
+2. > Let the event list be as meaningful as is necessary, and no more.
+
+The cartographer controller could emit the following events:
+
+| Reason | Message Format | Description | involvedObject |
+| --- | --- | --- | --- |
+| StampedObjectExternalSpecChange | an external actor changed the spec of `<group>.<kind>/<name>` | Our cache of the spec, and the spec we just generated match, but the API server has a different one. A lot of these is evidence of thrashing with external resources | Owner |
+| StampedObjectExternalSpecChange | an external actor changed the spec of resource: `<resource name>` for object `<group>.<kind>` | Our cache of the spec, and the spec we just generated match, but the API server has a different one. A lot of these is evidence of thrashing with external resources | Blueprint |
+| StampedObjectGetError | `<resource name>` could not retrieve `<group>.<kind>/<name>` due to error: `<error message>` | Loading a resource is failing due to a client.get issue (missing is not an error) | Owner |
+| StampedObjectObjectGetError | `<resource name>` could not retrieve `<group>.<kind>` due to error: `<error message>` | Loading a resources is failing due to a client.get issue (missing is not an error) - this could be spammy | Blueprint |
+| StampedObjectGarbageCollected | `<group>.<kind>/<name>` is no longer referenced | This owner has selected a different template, either by supply chain selection or templating, and this object is no longer needed | Owner |
+| ImmutableStampedObjectGarbageCollected | `<n> * <group>.<kind>` historical objects deleted due to garbage collection policy | This runnable's GC policy has caused `n` objects to be removed | Runnable |
+| StampedObjectInvalid | `<resource name>` could not be applied as `<group>.<kind>/<name>` due to API server error `<error message>` | This object was (probably) malformed. | Owner |
+| StampedObjectInvalid | `<resource name>` could not be applied as `<group>.<kind>` due to API server error `<error message>` | This object was (probably) malformed. This lets operators know their templates might have issues | Blueprint |
+| StampedObjectKindNotFound | `<resource name>` could not be applied because `<group>.<kind>/<name>` does not exist on this cluster | Did someone forget to install the CRDs? Otherwise it's a malformed template | Owner |
+| StampedObjectKindNotFound | `<resource name>` could not be applied because `<group>.<kind>` does not exist on this cluster | Did someone forget to install the CRDs? Otherwise it's a malformed template | Blueprint |
+| StampedObjectApplied | `<resource name>` was applied as `<group>.<kind>/<name>` | a stamped object needed to be created/updated | Owner |
```

Comment:
Do we want to distinguish between created and updated?

Reply:
Personally I think no. Compare separate:

```
Normal    StampedObjectCreated      1h (x1 over 1h)          workload-controller      `source-controller` was created as `sourcerepository.something/petclinic`
Normal    StampedObjectChanged     1h (x102 over 1h)     workload-controller      `source-controller` was updated as `sourcerepository.something/petclinic`
```

And combined:

```
Normal    StampedObjectApplied.      1h (x103 over 1h)     workload-controller      `source-controller` was applied as `sourcerepository.something/petclinic`
```

Essentially when separate, the creation event will be a singleton (except in rare cases when it was deleted, which is covered by two other events, one for deliberately by us, and the other for "spec change"). So we end up with a singleton event for all the resources, and then the combined applied events. All the singleton tells us is when it was created, which the `age` can do equally well. And if you dive into the event, it will tell you when the first event was pubaished _precisely_, so you can inspect the combined event and discover those details without needing a seperate event.